libcontainer/utils/cmsg.c

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2016 SUSE Linux GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "cmsg.h"
+
+#define error(fmt, ...)							\
+	({								\
+		fprintf(stderr, "nsenter: " fmt ": %m\n", ##__VA_ARGS__); \
+		errno = ECOMM;						\
+		-1; /* return value */					\
+	})
+
+/*
+ * Sends a file descriptor along the sockfd provided. Returns the return
+ * value of sendmsg(2). Any synchronisation and preparation of state
+ * should be done external to this (we expect the other side to be in
+ * recvfd() in the code).
+ */
+ssize_t sendfd(int sockfd, struct file_t file)
+{
+	struct msghdr msg = {0};
+	struct iovec iov[1] = {0};
+	struct cmsghdr *cmsg;
+	int *fdptr;
+	int ret;
+
+	union {
+		char buf[CMSG_SPACE(sizeof(file.fd))];
+		struct cmsghdr align;
+	} u;
+
+	/*
+	 * We need to send some other data along with ancilliary data, otherwise
+	 * the other side won't recieve any data. This is very well-hidden in the
+	 * documentation.
+	 */
+	iov[0].iov_base = file.tag;
+	iov[0].iov_len = strlen(file.tag) + 1;
```

Comment:
Sending a tag along with the file descriptor is a good idea for avoiding confusion (“here comes the pseudoterminal master”, `<fd>`).  But I didn't bother setting `msg.msg_iov` or `msg.msg_iovlen` in ccon (short of initializing `msg = {0}`), and I haven't had a problem without them.  The `SCM_RIGHTS` example in cmsg(3) doesn't bother with the `iov` fields either (although it doesn't explicitly call `sendmsg`).  Can you link to the docs you reference from your comment?